src/or/connection_edge.c

```diff
@@ -446,6 +454,172 @@ connection_edge_end(edge_connection_t *conn, uint8_t reason)
   return 0;
 }
 
+/**
+ * Add a half-closed connection to the list, to watch for activity.
+ *
+ * These connections are removed from the list upon receiving an end
+ * cell.
+ */
+static void
+connection_half_edge_add(edge_connection_t *conn,
+                         origin_circuit_t *circ)
+{
+  half_edge_t *half_conn = tor_malloc_zero(sizeof(half_edge_t));
+
+  if (!circ->half_streams) {
+    circ->half_streams = smartlist_new();
+  }
+
+  half_conn->stream_id = conn->stream_id;
+
+  // How many sendme's should I expect?
+  half_conn->sendmes_pending =
+   (STREAMWINDOW_START-conn->package_window)/STREAMWINDOW_INCREMENT;
+
+   // Is there a connected cell pending?
+  half_conn->connected_pending = conn->base_.state ==
+      AP_CONN_STATE_CONNECT_WAIT;
+
+  /* Data should only arrive if we're not waiting on a resolved.
+   * It can arrive after waiting on connected, because of optimistic
+   * data. */
+  if (conn->base_.state != AP_CONN_STATE_RESOLVE_WAIT) {
+    // How many more data cells can arrive on this id?
+    half_conn->data_pending = conn->deliver_window;
+  }
+
+  smartlist_add(circ->half_streams, half_conn);
+}
+
+/**
+ * Check if this stream_id is in a half-closed state. If so,
+ * check if it still has data cells pending, and decrement that
+ * window if so.
+ *
+ * Return 1 if the data window was not empty.
+ * Return 0 otherwise.
+ */
+int
+connection_half_edge_is_valid_data(smartlist_t *half_conns,
+                                   streamid_t stream_id)
+{
+  if (!half_conns)
+    return 0;
+
+  SMARTLIST_FOREACH_BEGIN(half_conns, half_edge_t *, half_conn) {
```

Comment:
The half_conns list should be quite small, and is checked infrequently. Connections only get added when a client hangs up on a TCP connection early, before getting the END cell from the other side. This typically only happens in cases where the user interrupts client action by a cancel/stop/kill at the application layer.

It would take a very broken client to fill the half_closed list, and even if the half_closed connection table manages to get full, it will be at most 65535 uint16_t equality comparisons to go through it. And it will only be checked when using that same circuit for a new stream, or when getting stray data back with stale stream ids on that circuit.

I actually think that the additional memory allocation, indirection, indirect calls, and hashing overhead involved in a hash table will make it considerably more expensive to use for these small sizes, especially since we're only comparing int16_t values in the search.

All of this made me opt for the simple list for this (which is also what we use for our opened streams). The simple list also felt easier to check for lifetime and correctness issues. (Though the refactoring you suggested does make this even easier).